{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "34cee08c-b6ee-450d-ae33-edd181df02b4",
   "metadata": {},
   "source": [
    "# Analyze the variant counts\n",
    "Analyze the counts of each variant in each sample as determined from the barcode sequencing.\n",
    "\n",
    "First, import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2b4fb9f-ad03-4348-8c99-1dfda78b0dde",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "\n",
    "import Bio.SeqIO\n",
    "\n",
    "import dms_variants.codonvarianttable\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfdbecf5-ead3-4d3e-a0ae-106cc124bce5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# allow more rows for Altair\n",
    "_ = alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "649e0dcc-7e23-4386-84d8-3cafba0ec69b",
   "metadata": {},
   "source": [
    "Read configuration:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aa10c89-ae3b-43da-a707-a6bef3d454d6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# If you are running notebook interactively rather than in pipeline that handles\n",
    "# working directories, you may have to first `os.chdir` to appropriate directory.\n",
    "\n",
    "with open(\"config.yaml\") as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "144fd6c2-0e57-4dae-91c3-87da91c62a0e",
   "metadata": {},
   "source": [
    "Read in all the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17189393-3282-4a26-abde-a47c4bdc755c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the barcode runs and get the samples\n",
    "barcode_runs = (\n",
    "    pd.read_csv(config[\"barcode_runs\"], parse_dates=[\"date\"])[\n",
    "        [\"sample\", \"library\", \"date\"]\n",
    "    ]\n",
    "    .sort_values([\"date\", \"sample\"])\n",
    "    .assign(date=lambda x: x[\"date\"].dt.strftime(\"%Y-%m-%d\"))\n",
    ")\n",
    "samples = barcode_runs[\"sample\"].tolist()\n",
    "assert len(samples) == len(set(samples))\n",
    "\n",
    "# read the barcode counts, invalid counts, and fates\n",
    "count_dfs = {}\n",
    "for count_type in [\"counts\", \"invalid\", \"fates\"]:\n",
    "    count_dfs[count_type] = pd.concat(\n",
    "        [\n",
    "            pd.read_csv(f\"results/barcode_counts/{sample}_{count_type}.csv\").assign(\n",
    "                sample=sample\n",
    "            )\n",
    "            for sample in samples\n",
    "        ]\n",
    "    )\n",
    "\n",
    "# read the site numbering map\n",
    "site_numbering_map = pd.read_csv(config[\"site_numbering_map\"])\n",
    "\n",
    "# create the codon-variant table\n",
    "variants = dms_variants.codonvarianttable.CodonVariantTable(\n",
    "    barcode_variant_file=config[\"codon_variants\"],\n",
    "    geneseq=str(Bio.SeqIO.read(config[\"gene_sequence_codon\"], \"fasta\").seq),\n",
    "    allowgaps=True,\n",
    "    substitutions_are_codon=True,\n",
    "    primary_target=\"gene\",\n",
    "    substitutions_col=\"codon_substitutions\",\n",
    ")\n",
    "variants.add_sample_counts_df(\n",
    "    count_dfs[\"counts\"].merge(\n",
    "        barcode_runs[[\"sample\", \"library\"]], on=\"sample\", validate=\"many_to_one\"\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d0d0c6b-34fe-404a-80df-d66d01e639da",
   "metadata": {},
   "source": [
    "## Barcode sequencing stats\n",
    "Plot \"fates\" of barcode reads: how many properly aligned to valid barcodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c68c09c0-8268-4f8a-8688-2f9c31aa7fe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "selections = [\n",
    "    alt.selection_point(\n",
    "        fields=[col],\n",
    "        bind=alt.binding_select(\n",
    "            options=[None] + barcode_runs[col].dropna().unique().tolist(),\n",
    "            labels=[\"all\"] + [str(x) for x in barcode_runs[col].dropna().unique()],\n",
    "            name=col,\n",
    "        ),\n",
    "    )\n",
    "    for col in [\"library\", \"date\"]\n",
    "]\n",
    "\n",
    "fate_chart = (\n",
    "    alt.Chart(count_dfs[\"fates\"])\n",
    "    .transform_lookup(\n",
    "        \"sample\",\n",
    "        alt.LookupData(barcode_runs, \"sample\", fields=[\"date\", \"library\"]),\n",
    "    )\n",
    "    .encode(\n",
    "        x=alt.X(\n",
    "            \"count\", title=\"barcode sequencing counts\", axis=alt.Axis(format=\".2g\")\n",
    "        ),\n",
    "        y=alt.Y(\"sample\", title=None, sort=samples),\n",
    "        color=alt.Color(\n",
    "            \"fate\",\n",
    "            scale=alt.Scale(reverse=True),\n",
    "        ),\n",
    "        order=alt.Order(\"fate\", sort=\"descending\"),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if c == \"count\" else c\n",
    "            for c in count_dfs[\"fates\"].columns.tolist() + [\"date:N\", \"library:N\"]\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar()\n",
    "    .properties(width=350, height=alt.Step(13))\n",
    "    .configure_axis(labelLimit=500)\n",
    ")\n",
    "\n",
    "for selection in selections:\n",
    "    fate_chart = fate_chart.add_params(selection).transform_filter(selection)\n",
    "\n",
    "display(fate_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fd22a96-5e84-43ef-b4ce-4f25d6784bfe",
   "metadata": {},
   "source": [
    "## Average valid and invalid counts per barcode\n",
    "Plot average valid and invalid counts per barcode.\n",
    "Invalid counts mean parseable barcodes that nonetheless don't map to our library's set of valid barcodes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "feae0184-90bb-4c22-80c8-3a2dddd6dc82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge valid and invalid counts\n",
    "counts = pd.concat(\n",
    "    [\n",
    "        count_dfs[\"counts\"].assign(valid=\"valid\"),\n",
    "        count_dfs[\"invalid\"].assign(valid=\"invalid\"),\n",
    "    ]\n",
    ")\n",
    "\n",
    "avg_counts = counts.groupby([\"sample\", \"valid\"], as_index=False).aggregate(\n",
    "    avg_count=pd.NamedAgg(\"count\", \"mean\")\n",
    ")\n",
    "\n",
    "valid_selection = alt.selection_point(fields=[\"valid\"], bind=\"legend\")\n",
    "\n",
    "avg_counts_chart = (\n",
    "    alt.Chart(avg_counts)\n",
    "    .transform_lookup(\n",
    "        \"sample\",\n",
    "        alt.LookupData(barcode_runs, \"sample\", fields=[\"date\", \"library\"]),\n",
    "    )\n",
    "    .encode(\n",
    "        x=alt.X(\"avg_count\", title=\"average counts per barcode\"),\n",
    "        y=alt.Y(\"sample\", title=None, sort=samples),\n",
    "        yOffset=\"valid\",\n",
    "        color=alt.Color(\n",
    "            \"valid\",\n",
    "            title=\"valid barcode\",\n",
    "            scale=alt.Scale(domain=avg_counts[\"valid\"].unique()),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if c == \"avg_count\" else c\n",
    "            for c in avg_counts.columns.tolist() + [\"library:N\", \"date:N\"]\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar()\n",
    "    .properties(width=200, height=alt.Step(15, **{\"for\": \"position\"}))\n",
    "    .configure_axis(labelLimit=500)\n",
    "    .add_params(*selections, valid_selection)\n",
    "    .transform_filter(valid_selection)\n",
    ")\n",
    "for selection in selections:\n",
    "    avg_counts_chart = avg_counts_chart.transform_filter(selection)\n",
    "\n",
    "avg_counts_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89be10b8-a198-4968-ac13-643439a57d66",
   "metadata": {},
   "source": [
    "## Check for abundant invalid barcodes\n",
    "Check for each sample to see if any invalid barcodes in the top most abundant.\n",
    "If so, that is a likely sign of contamination."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d50b0a0-4ef4-4e6e-88ad-688dd534d66b",
   "metadata": {},
   "outputs": [],
   "source": [
    "invalid_rank = counts.assign(\n",
    "    rank=lambda x: x.groupby(\"sample\")[\"count\"].rank(method=\"first\")\n",
    ")\n",
    "\n",
    "invalid_n = 20  # look for invalid in top this many\n",
    "\n",
    "invalid_top_n_counts = (\n",
    "    invalid_rank.sort_values(\"rank\")\n",
    "    .groupby([\"sample\", \"valid\"], as_index=False)\n",
    "    .aggregate(\n",
    "        invalid_in_top_n=pd.NamedAgg(\"rank\", lambda s: len(s[s <= invalid_n])),\n",
    "        top_invalid_rank=pd.NamedAgg(\"rank\", \"first\"),\n",
    "        top_invalid_barcode=pd.NamedAgg(\"barcode\", \"first\"),\n",
    "    )\n",
    "    .query(\"valid == 'invalid'\")\n",
    ")\n",
    "\n",
    "invalid_top_n_counts_chart = (\n",
    "    alt.Chart(invalid_top_n_counts)\n",
    "    .transform_lookup(\n",
    "        \"sample\",\n",
    "        alt.LookupData(barcode_runs, \"sample\", fields=[\"date\", \"library\"]),\n",
    "    )\n",
    "    .encode(\n",
    "        x=alt.X(\n",
    "            \"invalid_in_top_n\",\n",
    "            title=f\"invalid barcodes in top {invalid_n}\",\n",
    "            scale=alt.Scale(domain=(0, invalid_n)),\n",
    "        ),\n",
    "        y=alt.Y(\"sample\", title=None, sort=samples),\n",
    "        tooltip=invalid_top_n_counts.columns.tolist(),\n",
    "    )\n",
    "    .mark_bar()\n",
    "    .properties(width=200, height=alt.Step(15))\n",
    "    .configure_axis(labelLimit=500)\n",
    "    .add_params(*selections)\n",
    ")\n",
    "for selection in selections:\n",
    "    invalid_top_n_counts_chart = invalid_top_n_counts_chart.transform_filter(selection)\n",
    "\n",
    "display(invalid_top_n_counts_chart)\n",
    "\n",
    "print(\"\\nHere are top invalid barcodes:\")\n",
    "invalid_rank.query(\"valid == 'invalid'\").query(\"rank <= @invalid_n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de188539-dffb-413f-aa9f-709812c6bd0b",
   "metadata": {},
   "source": [
    "## Counts of each target type\n",
    "We may have different target types in addition to the gene, such as a spike-in or neutralization standard.\n",
    "Plot the average counts per variant of each target type:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52be6393-37b9-468c-8b8a-04ad2c9180e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "avg_counts_by_target = variants.avgCountsPerVariant(libraries=variants.libraries).merge(\n",
    "    barcode_runs, validate=\"many_to_one\"\n",
    ")\n",
    "\n",
    "target_selection = alt.selection_point(fields=[\"target\"], bind=\"legend\", value=\"gene\")\n",
    "\n",
    "avg_counts_by_target_chart = (\n",
    "    alt.Chart(avg_counts_by_target)\n",
    "    .encode(\n",
    "        x=alt.X(\"avg_counts_per_variant\", title=\"average counts per variant\"),\n",
    "        y=alt.Y(\"sample\", title=None, sort=samples),\n",
    "        yOffset=\"target\",\n",
    "        color=alt.Color(\n",
    "            \"target\",\n",
    "            title=\"variant type\",\n",
    "            scale=alt.Scale(domain=avg_counts_by_target[\"target\"].unique()),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if c == \"avg_count_per_variant\" else c\n",
    "            for c in avg_counts_by_target.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar()\n",
    "    .properties(width=200, height=alt.Step(15, **{\"for\": \"position\"}))\n",
    "    .configure_axis(labelLimit=500)\n",
    "    .add_params(*selections, target_selection)\n",
    "    .transform_filter(target_selection)\n",
    ")\n",
    "for selection in selections:\n",
    "    avg_counts_by_target_chart = avg_counts_by_target_chart.transform_filter(selection)\n",
    "\n",
    "avg_counts_by_target_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "926aa73b-f54a-447d-a64c-b1a44b3b9248",
   "metadata": {},
   "source": [
    "Plot fraction of all counts that are to each type of target:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52619bb8-02ee-4f7c-9307-891daabf0425",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_fracs = (\n",
    "    variants.variant_count_df.groupby([\"sample\", \"library\", \"target\"])\n",
    "    .aggregate(total_counts=pd.NamedAgg(\"count\", \"sum\"))\n",
    "    .assign(\n",
    "        frac_counts=lambda x: (\n",
    "            x[\"total_counts\"]\n",
    "            / x.groupby([\"sample\", \"library\"])[\"total_counts\"].transform(\"sum\")\n",
    "        )\n",
    "    )\n",
    "    .drop(columns=\"total_counts\")\n",
    "    .reset_index()\n",
    "    .merge(barcode_runs, validate=\"many_to_one\")\n",
    ")\n",
    "\n",
    "target_selection = alt.selection_point(fields=[\"target\"], bind=\"legend\")\n",
    "\n",
    "target_fracs_chart = (\n",
    "    alt.Chart(target_fracs)\n",
    "    .encode(\n",
    "        x=alt.X(\"frac_counts\", title=\"fraction of variant counts\"),\n",
    "        y=alt.Y(\"sample\", title=None, sort=samples),\n",
    "        color=alt.Color(\n",
    "            \"target\",\n",
    "            title=\"variant type\",\n",
    "            scale=alt.Scale(domain=target_fracs[\"target\"].unique()),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if c == \"frac_counts\" else c\n",
    "            for c in target_fracs.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar()\n",
    "    .properties(width=200, height=alt.Step(12))\n",
    "    .configure_axis(labelLimit=500)\n",
    "    .add_params(*selections, target_selection)\n",
    "    .transform_filter(target_selection)\n",
    ")\n",
    "for selection in selections:\n",
    "    target_fracs_chart = target_fracs_chart.transform_filter(selection)\n",
    "\n",
    "target_fracs_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "089b65f0-8b06-4627-8e3e-85130f548da1",
   "metadata": {},
   "source": [
    "## Diversity of barcodes for each sample\n",
    "Get the diversity of barcodes mapping to the gene of interest (**not** neutralization standards or other non-primary targets) for each sample.\n",
    "\n",
    "We compute three [diversity indices](https://en.wikipedia.org/wiki/Diversity_index):\n",
    "\n",
    "  - $^0$D: number ofo unique barcodes observed in that sample\n",
    "  - $^1$D: effective number of barcodes (Shannon diversity)\n",
    "  - $^2$D: inverse Simpson index (inverse probability two random barcodes are the same)\n",
    "\n",
    "All are plotted below, the plot is interactive, so you can select by date library and click the legend just to show on particular diversity measure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48365603-3de6-4016-9e67-dceb6088f441",
   "metadata": {},
   "outputs": [],
   "source": [
    "diversity_measures = {\n",
    "    \"unique_barcodes_observed\": \"0D (unique barcodes observed)\",\n",
    "    \"effective_n_barcodes\": \"1D (effective barcodes)\",\n",
    "    \"inverse_simpson_index\": \"2D (inverse Simpson index)\",\n",
    "}\n",
    "\n",
    "diversity = (\n",
    "    variants.variant_count_df.query(\"target == @variants.primary_target\")\n",
    "    .query(\"count > 0\")[[\"sample\", \"barcode\", \"count\"]]\n",
    "    .groupby(\"sample\", as_index=False)\n",
    "    .aggregate(\n",
    "        unique_barcodes_observed=pd.NamedAgg(\"barcode\", \"count\"),\n",
    "        effective_n_barcodes=pd.NamedAgg(\n",
    "            \"count\",\n",
    "            lambda s: numpy.exp(-((s / s.sum()) * numpy.log(s / s.sum())).sum()),\n",
    "        ),\n",
    "        inverse_simpson_index=pd.NamedAgg(\n",
    "            \"count\",\n",
    "            lambda s: 1 / ((s / s.sum()) ** 2).sum(),\n",
    "        ),\n",
    "    )\n",
    "    .melt(id_vars=\"sample\", var_name=\"diversity_measure\", value_name=\"diversity\")\n",
    "    .merge(barcode_runs, on=\"sample\", validate=\"many_to_one\")\n",
    "    .assign(diversity_measure=lambda x: x[\"diversity_measure\"].map(diversity_measures))\n",
    ")\n",
    "\n",
    "diversity_selection = alt.selection_point(fields=[\"diversity_measure\"], bind=\"legend\")\n",
    "\n",
    "diversity_chart = (\n",
    "    alt.Chart(diversity)\n",
    "    .encode(\n",
    "        x=alt.X(\"diversity\"),\n",
    "        y=alt.Y(\"sample\", title=None, sort=samples),\n",
    "        yOffset=\"diversity_measure\",\n",
    "        color=alt.Color(\n",
    "            \"diversity_measure\",\n",
    "            scale=alt.Scale(domain=list(diversity_measures.values())),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".1f\") if diversity[c].dtype == float else c\n",
    "            for c in diversity.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar()\n",
    "    .properties(width=200, height=alt.Step(15, **{\"for\": \"position\"}))\n",
    "    .configure_axis(labelLimit=500)\n",
    ")\n",
    "\n",
    "for selection in selections + [diversity_selection]:\n",
    "    diversity_chart = diversity_chart.add_params(selection).transform_filter(selection)\n",
    "\n",
    "diversity_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ef93ec5-52ce-40f2-8380-d7d9ac45892b",
   "metadata": {},
   "source": [
    "## Average mutations per variant\n",
    "Plot average number of mutations per variant of each type of codon mutation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a16013d4-fccd-4c40-8ff8-506deb6f69ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "avg_muts = (\n",
    "    variants.numCodonMutsByType(\n",
    "        variant_type=\"all\",\n",
    "        libraries=variants.libraries,\n",
    "    )\n",
    "    .drop(columns=[\"count\", \"num_muts_count\"])\n",
    "    .merge(barcode_runs, validate=\"many_to_one\")\n",
    "    # remove categorical to enable plotting below\n",
    "    .assign(mutation_type=lambda x: x[\"mutation_type\"].tolist())\n",
    "    .rename(columns={\"number\": \"average mutations\"})\n",
    ")\n",
    "\n",
    "mut_type_selection = alt.selection_point(fields=[\"mutation_type\"], bind=\"legend\")\n",
    "mut_types = avg_muts[\"mutation_type\"].unique()\n",
    "\n",
    "avg_muts_chart = (\n",
    "    alt.Chart(avg_muts)\n",
    "    .encode(\n",
    "        x=alt.X(\"average mutations\", axis=alt.Axis(grid=False)),\n",
    "        y=alt.Y(\"sample\", title=None),\n",
    "        yOffset=alt.YOffset(\"mutation_type\", sort=mut_types),\n",
    "        color=alt.Color(\"mutation_type\", scale=alt.Scale(domain=mut_types)),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if c == \"average mutations\" else c\n",
    "            for c in avg_muts.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_bar()\n",
    "    .properties(height=alt.Step(20, **{\"for\": \"position\"}), width=300)\n",
    "    .configure_axis(labelLimit=500)\n",
    "    .add_params(mut_type_selection, *selections)\n",
    "    .transform_filter(mut_type_selection)\n",
    ")\n",
    "\n",
    "for selection in selections:\n",
    "    avg_muts_chart = avg_muts_chart.transform_filter(selection)\n",
    "\n",
    "avg_muts_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b32567c-e689-49c4-aca0-dc21374708d8",
   "metadata": {},
   "source": [
    "## Mutation frequencies across gene\n",
    "Plot percent of all variants with any type of amino-acid changing mutation type (nonsynonymous, stop, or deletion) at each site:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7371df2-a11a-4ab7-a5a7-37f009f7c0de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get site mutation frequencies in wide form to make it as compact as possible.\n",
    "# Reason is this is a lot of data and storing compactly shrinks plot.\n",
    "# We also encode the samples as integers because we can't transform_fold them\n",
    "# if strings containing a dot\n",
    "sample_coding = {sample: str(i) for (i, sample) in enumerate(samples)}\n",
    "\n",
    "site_freqs_wide = (\n",
    "    variants.mutCounts(variant_type=\"all\", mut_type=\"aa\", libraries=variants.libraries)\n",
    "    .groupby([\"sample\", \"site\"])\n",
    "    .aggregate(mutation_count=pd.NamedAgg(\"count\", \"sum\"))\n",
    "    .reset_index()\n",
    "    .merge(\n",
    "        variants.n_variants_df(libraries=variants.libraries, primary_target_only=True),\n",
    "        validate=\"many_to_one\",\n",
    "    )\n",
    "    .assign(percent_w_mutation=lambda x: 100 * x[\"mutation_count\"] / x[\"count\"])\n",
    "    .pivot_table(\n",
    "        index=\"site\",\n",
    "        values=\"percent_w_mutation\",\n",
    "        columns=\"sample\",\n",
    "    )\n",
    "    .reset_index()\n",
    "    .rename(columns={\"site\": \"sequential_site\"})\n",
    "    .rename(columns=sample_coding)\n",
    "    .assign(wildtype=lambda x: x[\"sequential_site\"].map(variants.aas))\n",
    "    .merge(\n",
    "        site_numbering_map[[\"sequential_site\", \"reference_site\"]],\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    ")\n",
    "\n",
    "assert not site_freqs_wide.isnull().any().any()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03aec811-d278-428b-a3fd-8e2d4e73e2af",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-26T23:31:08.043701Z",
     "iopub.status.busy": "2023-06-26T23:31:08.043372Z",
     "iopub.status.idle": "2023-06-26T23:31:08.049161Z",
     "shell.execute_reply": "2023-06-26T23:31:08.048633Z",
     "shell.execute_reply.started": "2023-06-26T23:31:08.043679Z"
    }
   },
   "source": [
    "Make the plot.\n",
    "Note that you can select which libraries or dates to show with the dropdown box at bottom:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "320bdab0-067a-4a51-9f3f-21e97182cd27",
   "metadata": {},
   "outputs": [],
   "source": [
    "site_freqs_chart = (\n",
    "    alt.Chart(site_freqs_wide)\n",
    "    .transform_fold(list(sample_coding.values()), [\"sample_coding\", \"percent\"])\n",
    "    .transform_lookup(\n",
    "        lookup=\"sample_coding\",\n",
    "        from_=alt.LookupData(\n",
    "            data=barcode_runs.assign(\n",
    "                sample_coding=lambda x: x[\"sample\"].map(sample_coding)\n",
    "            ),\n",
    "            key=\"sample_coding\",\n",
    "            fields=[\"library\", \"date\", \"sample\"],\n",
    "        ),\n",
    "    )\n",
    "    .encode(\n",
    "        x=alt.X(\n",
    "            \"reference_site\",\n",
    "            scale=alt.Scale(nice=False, zero=False),\n",
    "            sort=alt.SortField(\"sequential_site\"),\n",
    "        ),\n",
    "        y=alt.Y(\"percent:Q\", title=\"% variants w mutation\"),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if c in {\"percent:Q\"} else c\n",
    "            for c in [\n",
    "                \"sequential_site\",\n",
    "                \"reference_site\",\n",
    "                \"wildtype\",\n",
    "                \"percent:Q\",\n",
    "                \"date:N\",\n",
    "            ]\n",
    "        ],\n",
    "        facet=alt.Facet(\n",
    "            \"sample:N\",\n",
    "            title=None,\n",
    "            header=alt.Header(labelPadding=0),\n",
    "            columns=3,\n",
    "            spacing=5,\n",
    "        ),\n",
    "    )\n",
    "    .mark_line(point=True, size=0.5)\n",
    "    .properties(height=100, width=275, autosize=alt.AutoSizeParams(resize=True))\n",
    "    .add_params(*selections)\n",
    "    .resolve_scale(y=\"independent\")\n",
    "    .configure_axis(grid=False)\n",
    ")\n",
    "\n",
    "for selection in selections:\n",
    "    site_freqs_chart = site_freqs_chart.transform_filter(selection)\n",
    "\n",
    "site_freqs_chart"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
