<h1 id="deep-mutational-scanning-dms-pipeline-for-a-viral-entry-protein-vep">Deep mutational scanning (DMS) pipeline for a viral entry protein (VEP)</h1>
<p>Analysis by Jesse Bloom (2023)</p>
<p>See <a href="https://github.com/dms-vep/dms-vep-pipeline-3">https://github.com/dms-vep/dms-vep-pipeline-3</a> for full code.</p>
<div class="toc"><span class="toctitle">Contents</span><ul>
<li><a href="#barcode-to-codon-variant-lookup-table">Barcode to codon-variant lookup table</a></li>
<li><a href="#count-barcodes-for-variants">Count barcodes for variants</a></li>
<li><a href="#functional-effects-of-mutations">Functional effects of mutations</a></li>
<li><a href="#antibodyserum-escape">Antibody/serum escape</a></li>
<li><a href="#receptor-affinity">Receptor affinity</a></li>
<li><a href="#integrated-summary-plots">Integrated summary plots</a></li>
<li><a href="#additional-analysis-specific-files">Additional analysis-specific files</a></li>
</ul>
</div>
<h2 id="barcode-to-codon-variant-lookup-table">Barcode to codon-variant lookup table</h2>
<h4 id="analysis-notebooks">Analysis notebooks</h4>
<ul>
<li><a href="notebooks/analyze_pacbio_ccs.html">Analysis of PacBio CCSs</a></li>
<li><a href="notebooks/build_pacbio_consensus.html">Building barcoded variant consensus sequences</a></li>
<li><a href="notebooks/build_codon_variants.html">Building of codon-variant table</a></li>
</ul>
<h4 id="data-files">Data files</h4>
<ul>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/gene_sequence/codon.fasta">parental codon sequence</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/gene_sequence/protein.fasta">parental protein sequence</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/variants/codon_variants.csv">barcode to codon-variant lookup table</a></li>
</ul>
<h2 id="count-barcodes-for-variants">Count barcodes for variants</h2>
<h4 id="analysis-notebooks_1">Analysis notebooks</h4>
<ul>
<li><a href="notebooks/analyze_variant_counts.html">Analysis of variant counts</a></li>
</ul>
<h4 id="data-files_1">Data files</h4>
<ul>
<li><details><summary>Barcode count CSVs (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220210-REGN10933-0.15-1_counts.csv">LibA-220210-REGN10933-0.15-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220210-REGN10933-1.39-1_counts.csv">LibA-220210-REGN10933-1.39-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220210-REGN10933-5.58-1_counts.csv">LibA-220210-REGN10933-5.58-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220210-no_antibody-1_counts.csv">LibA-220210-no_antibody-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220210-REGN10933-0.15-2_counts.csv">LibA-220210-REGN10933-0.15-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220210-REGN10933-1.39-2_counts.csv">LibA-220210-REGN10933-1.39-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220210-REGN10933-5.58-2_counts.csv">LibA-220210-REGN10933-5.58-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220210-no_antibody-2_counts.csv">LibA-220210-no_antibody-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibB-220302-REGN10933-0.15-1_counts.csv">LibB-220302-REGN10933-0.15-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibB-220302-REGN10933-1.39-1_counts.csv">LibB-220302-REGN10933-1.39-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibB-220302-REGN10933-5.58-1_counts.csv">LibB-220302-REGN10933-5.58-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibB-220302-no_antibody-1_counts.csv">LibB-220302-no_antibody-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220302-S2M11-0.3082-1_counts.csv">LibA-220302-S2M11-0.3082-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220302-S2M11-1.2328-1_counts.csv">LibA-220302-S2M11-1.2328-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220302-S2M11-4.9314-1_counts.csv">LibA-220302-S2M11-4.9314-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220302-S2M11-0.3082-2_counts.csv">LibA-220302-S2M11-0.3082-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220302-S2M11-1.2328-2_counts.csv">LibA-220302-S2M11-1.2328-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220302-S2M11-4.9314-2_counts.csv">LibA-220302-S2M11-4.9314-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220302-no_antibody-2_counts.csv">LibA-220302-no_antibody-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-220302-no_antibody-1_counts.csv">LibA-220302-no_antibody-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-211028-VSVG-1_counts.csv">LibA-211028-VSVG-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibA-211028-VSVG-2_counts.csv">LibA-211028-VSVG-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/barcode_counts/LibB-211028-VSVG-1_counts.csv">LibB-211028-VSVG-1</a></li>
</ul></details>

</li>
</ul>
<h2 id="functional-effects-of-mutations">Functional effects of mutations</h2>
<h4 id="final-summary-plots">Final summary plots</h4>
<ul>
<li><details><summary>Interactive plots of average mutation functional effects (click triangle to expand/collapse)</summary><ul>

<li><a href="htmls/293T_ACE2_entry_func_effects.html">293T_ACE2_entry</a></li>
</ul></details>

</li>
<li><details><summary>Interactive plots of average mutation latent-phenotype effects (click triangle to expand/collapse)</summary><ul>

<li><a href="htmls/293T_ACE2_entry_latent_effects.html">293T_ACE2_entry</a></li>
</ul></details>

</li>
<li><details><summary>Interactive plots of differences in functional effects between conditions (click triangle to expand/collapse)</summary><ul>

<li><a href="htmls/220210_vs_220302_comparison_diffs.html">220210_vs_220302_comparison</a></li>
</ul></details>

</li>
<li><details><summary>Interactive plots of average shifts in functional effects (click triangle to expand/collapse)</summary><ul>

<li><a href="htmls/date_comparison_shifts.html">date_comparison</a></li>
</ul></details>

</li>
</ul>
<h4 id="analysis-notebooks_2">Analysis notebooks</h4>
<ul>
<li><a href="notebooks/analyze_func_scores.html">Analysis of functional scores</a></li>
<li><details><summary>Notebooks with per-selection global epistasis fitting (click triangle to expand/collapse)</summary><ul>

<li><a href="notebooks/func_effects_global_epistasis_LibA-220210-293T_ACE2-1.html">LibA-220210-293T_ACE2-1</a></li>
<li><a href="notebooks/func_effects_global_epistasis_LibA-220210-293T_ACE2-2.html">LibA-220210-293T_ACE2-2</a></li>
<li><a href="notebooks/func_effects_global_epistasis_LibA-220302-293T_ACE2-1.html">LibA-220302-293T_ACE2-1</a></li>
<li><a href="notebooks/func_effects_global_epistasis_LibA-220302-293T_ACE2-2.html">LibA-220302-293T_ACE2-2</a></li>
<li><a href="notebooks/func_effects_global_epistasis_LibB-220302-293T_ACE2-1.html">LibB-220302-293T_ACE2-1</a></li>
</ul></details>

</li>
<li><details><summary>Notebooks averaging mutation functional effects (click triangle to expand/collapse)</summary><ul>

<li><a href="notebooks/avg_func_effects_293T_ACE2_entry.html">293T_ACE2_entry</a></li>
</ul></details>

</li>
<li><details><summary>Notebooks taking differences in functional effects between conditions (click triangle to expand/collapse)</summary><ul>

<li><a href="notebooks/func_effect_diffs_220210_vs_220302_comparison.html">220210_vs_220302_comparison</a></li>
</ul></details>

</li>
<li><details><summary>Notebooks fitting shifts in functional effects (click triangle to expand/collapse)</summary><ul>

<li><a href="notebooks/func_effect_shifts_LibA-date_comparison-1.html">LibA-date_comparison-1</a></li>
<li><a href="notebooks/func_effect_shifts_LibA-date_comparison-2.html">LibA-date_comparison-2</a></li>
</ul></details>

</li>
<li><details><summary>Notebooks averaging shifts in functional effects (click triangle to expand/collapse)</summary><ul>

<li><a href="notebooks/avg_func_effect_shifts_date_comparison.html">date_comparison</a></li>
</ul></details>

</li>
</ul>
<h4 id="data-files_2">Data files</h4>
<ul>
<li><details><summary>Per-variant functional score CSVs (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_scores/LibA-220210-293T_ACE2-1_func_scores.csv">LibA-220210-293T_ACE2-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_scores/LibA-220210-293T_ACE2-2_func_scores.csv">LibA-220210-293T_ACE2-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_scores/LibA-220302-293T_ACE2-1_func_scores.csv">LibA-220302-293T_ACE2-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_scores/LibA-220302-293T_ACE2-2_func_scores.csv">LibA-220302-293T_ACE2-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_scores/LibB-220302-293T_ACE2-1_func_scores.csv">LibB-220302-293T_ACE2-1</a></li>
</ul></details>

</li>
<li><details><summary>Per-selection mutation functional effect CSVs (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_effects/by_selection/LibA-220210-293T_ACE2-1_func_effects.csv">LibA-220210-293T_ACE2-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_effects/by_selection/LibA-220210-293T_ACE2-2_func_effects.csv">LibA-220210-293T_ACE2-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_effects/by_selection/LibA-220302-293T_ACE2-1_func_effects.csv">LibA-220302-293T_ACE2-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_effects/by_selection/LibA-220302-293T_ACE2-2_func_effects.csv">LibA-220302-293T_ACE2-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_effects/by_selection/LibB-220302-293T_ACE2-1_func_effects.csv">LibB-220302-293T_ACE2-1</a></li>
</ul></details>

</li>
<li><details><summary>Average mutation functional effects (CSV files) (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_effects/averages/293T_ACE2_entry_func_effects.csv">293T_ACE2_entry</a></li>
</ul></details>

</li>
<li><details><summary>Average mutation latent-phenotype effects (CSV files) (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_effects/averages/293T_ACE2_entry_func_effects.csv">293T_ACE2_entry</a></li>
</ul></details>

</li>
<li><details><summary>Differences in functional effects (CSV files) (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_effect_diffs/220210_vs_220302_comparison_diffs.csv">220210_vs_220302_comparison</a></li>
</ul></details>

</li>
<li><details><summary>Per-condition functional effect shifts CSVs (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_effect_shifts/by_comparison/LibA-date_comparison-1_shifts.csv">LibA-date_comparison-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_effect_shifts/by_comparison/LibA-date_comparison-2_shifts.csv">LibA-date_comparison-2</a></li>
</ul></details>

</li>
<li><details><summary>Average shifts in functional effects (CSV files) (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/func_effect_shifts/averages/date_comparison_shifts.csv">date_comparison</a></li>
</ul></details>

</li>
</ul>
<h2 id="antibodyserum-escape">Antibody/serum escape</h2>
<h4 id="final-summary-plots_1">Final summary plots</h4>
<ul>
<li><details><summary>antibody escape mutation effect plots (click triangle to expand/collapse)</summary><ul>

<li><a href="htmls/REGN10933_mut_effect.html">REGN10933</a></li>
<li><a href="htmls/S2M11_mut_effect.html">S2M11</a></li>
</ul></details>

</li>
</ul>
<h4 id="analysis-notebooks_3">Analysis notebooks</h4>
<ul>
<li><details><summary>Fits of polyclonal models to individual antibody escape selections (click triangle to expand/collapse)</summary><ul>

<li><a href="notebooks/fit_escape_antibody_escape_LibA-220210-REGN10933-1.html">LibA-220210-REGN10933-1</a></li>
<li><a href="notebooks/fit_escape_antibody_escape_LibA-220210-REGN10933-2.html">LibA-220210-REGN10933-2</a></li>
<li><a href="notebooks/fit_escape_antibody_escape_LibB-220302-REGN10933-1.html">LibB-220302-REGN10933-1</a></li>
<li><a href="notebooks/fit_escape_antibody_escape_LibA-220302-S2M11-1.html">LibA-220302-S2M11-1</a></li>
<li><a href="notebooks/fit_escape_antibody_escape_LibA-220302-S2M11-2.html">LibA-220302-S2M11-2</a></li>
</ul></details>

</li>
<li><details><summary>Average selections for antibody escape (click triangle to expand/collapse)</summary><ul>

<li><a href="notebooks/avg_escape_antibody_escape_REGN10933.html">REGN10933</a></li>
<li><a href="notebooks/avg_escape_antibody_escape_S2M11.html">S2M11</a></li>
</ul></details>

</li>
</ul>
<h4 id="data-files_3">Data files</h4>
<ul>
<li><details><summary>Probability (fraction) escape for each variant in each antibody escape selection (CSVs) (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibA-220210-REGN10933-1/LibA-220210-REGN10933-0.15-1_prob_escape.csv">LibA-220210-REGN10933-1 LibA-220210-REGN10933-0.15-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibA-220210-REGN10933-1/LibA-220210-REGN10933-1.39-1_prob_escape.csv">LibA-220210-REGN10933-1 LibA-220210-REGN10933-1.39-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibA-220210-REGN10933-1/LibA-220210-REGN10933-5.58-1_prob_escape.csv">LibA-220210-REGN10933-1 LibA-220210-REGN10933-5.58-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibA-220210-REGN10933-2/LibA-220210-REGN10933-0.15-2_prob_escape.csv">LibA-220210-REGN10933-2 LibA-220210-REGN10933-0.15-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibA-220210-REGN10933-2/LibA-220210-REGN10933-1.39-2_prob_escape.csv">LibA-220210-REGN10933-2 LibA-220210-REGN10933-1.39-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibA-220210-REGN10933-2/LibA-220210-REGN10933-5.58-2_prob_escape.csv">LibA-220210-REGN10933-2 LibA-220210-REGN10933-5.58-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibB-220302-REGN10933-1/LibB-220302-REGN10933-0.15-1_prob_escape.csv">LibB-220302-REGN10933-1 LibB-220302-REGN10933-0.15-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibB-220302-REGN10933-1/LibB-220302-REGN10933-1.39-1_prob_escape.csv">LibB-220302-REGN10933-1 LibB-220302-REGN10933-1.39-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibB-220302-REGN10933-1/LibB-220302-REGN10933-5.58-1_prob_escape.csv">LibB-220302-REGN10933-1 LibB-220302-REGN10933-5.58-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibA-220302-S2M11-1/LibA-220302-S2M11-0.3082-1_prob_escape.csv">LibA-220302-S2M11-1 LibA-220302-S2M11-0.3082-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibA-220302-S2M11-1/LibA-220302-S2M11-1.2328-1_prob_escape.csv">LibA-220302-S2M11-1 LibA-220302-S2M11-1.2328-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibA-220302-S2M11-1/LibA-220302-S2M11-4.9314-1_prob_escape.csv">LibA-220302-S2M11-1 LibA-220302-S2M11-4.9314-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibA-220302-S2M11-2/LibA-220302-S2M11-0.3082-2_prob_escape.csv">LibA-220302-S2M11-2 LibA-220302-S2M11-0.3082-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibA-220302-S2M11-2/LibA-220302-S2M11-1.2328-2_prob_escape.csv">LibA-220302-S2M11-2 LibA-220302-S2M11-1.2328-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/by_selection/LibA-220302-S2M11-2/LibA-220302-S2M11-4.9314-2_prob_escape.csv">LibA-220302-S2M11-2 LibA-220302-S2M11-4.9314-2</a></li>
</ul></details>

</li>
<li><details><summary>antibody escape CSVs (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/averages/REGN10933_mut_effect.csv">REGN10933</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/antibody_escape/averages/S2M11_mut_effect.csv">S2M11</a></li>
</ul></details>

</li>
</ul>
<h2 id="receptor-affinity">Receptor affinity</h2>
<h4 id="final-summary-plots_2">Final summary plots</h4>
<ul>
<li><details><summary>receptor affinity mutation effect plots (click triangle to expand/collapse)</summary><ul>

<li><a href="htmls/pretending_S2M11_is_receptor_mut_effect.html">pretending_S2M11_is_receptor</a></li>
</ul></details>

</li>
</ul>
<h4 id="analysis-notebooks_4">Analysis notebooks</h4>
<ul>
<li><details><summary>Fits of polyclonal models to individual receptor affinity selections (click triangle to expand/collapse)</summary><ul>

<li><a href="notebooks/fit_escape_receptor_affinity_LibA-220302-pretending_S2M11_is_receptor-1.html">LibA-220302-pretending_S2M11_is_receptor-1</a></li>
<li><a href="notebooks/fit_escape_receptor_affinity_LibA-220302-pretending_S2M11_is_receptor-2.html">LibA-220302-pretending_S2M11_is_receptor-2</a></li>
</ul></details>

</li>
<li><details><summary>Average selections for receptor affinity (click triangle to expand/collapse)</summary><ul>

<li><a href="notebooks/avg_escape_receptor_affinity_pretending_S2M11_is_receptor.html">pretending_S2M11_is_receptor</a></li>
</ul></details>

</li>
</ul>
<h4 id="data-files_4">Data files</h4>
<ul>
<li><details><summary>Probability (fraction) escape for each variant in each receptor affinity selection (CSVs) (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/receptor_affinity/by_selection/LibA-220302-pretending_S2M11_is_receptor-1/LibA-220302-S2M11-0.3082-1_prob_escape.csv">LibA-220302-pretending_S2M11_is_receptor-1 LibA-220302-S2M11-0.3082-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/receptor_affinity/by_selection/LibA-220302-pretending_S2M11_is_receptor-1/LibA-220302-S2M11-1.2328-1_prob_escape.csv">LibA-220302-pretending_S2M11_is_receptor-1 LibA-220302-S2M11-1.2328-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/receptor_affinity/by_selection/LibA-220302-pretending_S2M11_is_receptor-1/LibA-220302-S2M11-4.9314-1_prob_escape.csv">LibA-220302-pretending_S2M11_is_receptor-1 LibA-220302-S2M11-4.9314-1</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/receptor_affinity/by_selection/LibA-220302-pretending_S2M11_is_receptor-2/LibA-220302-S2M11-0.3082-2_prob_escape.csv">LibA-220302-pretending_S2M11_is_receptor-2 LibA-220302-S2M11-0.3082-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/receptor_affinity/by_selection/LibA-220302-pretending_S2M11_is_receptor-2/LibA-220302-S2M11-1.2328-2_prob_escape.csv">LibA-220302-pretending_S2M11_is_receptor-2 LibA-220302-S2M11-1.2328-2</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/receptor_affinity/by_selection/LibA-220302-pretending_S2M11_is_receptor-2/LibA-220302-S2M11-4.9314-2_prob_escape.csv">LibA-220302-pretending_S2M11_is_receptor-2 LibA-220302-S2M11-4.9314-2</a></li>
</ul></details>

</li>
<li><details><summary>receptor affinity CSVs (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/receptor_affinity/averages/pretending_S2M11_is_receptor_mut_effect.csv">pretending_S2M11_is_receptor</a></li>
</ul></details>

</li>
</ul>
<h2 id="integrated-summary-plots">Integrated summary plots</h2>
<h4 id="summary-of-all-phenotypes">summary of all phenotypes</h4>
<ul>
<li><a href="htmls/summary_of_all_phenotypes_overlaid.html">Summary plot (escape overlaid)</a></li>
<li><a href="htmls/summary_of_all_phenotypes_faceted.html">Summary plot (escape faceted)</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/summaries/summary_of_all_phenotypes.csv">CSV summarizing results</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/summaries/summary_of_all_phenotypes_per_antibody_escape.csv">CSV summarizing per-antibody results</a></li>
</ul>
<h4 id="summary-of-just-entry-and-binding">summary of just entry and binding</h4>
<ul>
<li><a href="htmls/summary_of_just_entry_and_binding_overlaid.html">Summary plot</a></li>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/results/summaries/summary_of_just_entry_and_binding.csv">CSV summarizing results</a></li>
</ul>
<h2 id="additional-analysis-specific-files">Additional analysis-specific files</h2>
<h4 id="data-files_5">Data files</h4>
<ul>
<li><a href="https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/data/site_numbering_map.csv">Reference to sequential site-numbering map</a></li>
</ul>